File: tools/hygiene/fix-markdown-md032-md026.py
#!/usr/bin/env python3
"""
tools/hygiene/fix-markdown-md032-md026.py — mechanical fix for two
markdownlint violations:

- MD032 (blanks-around-lists): inserts blank lines before/after list
  blocks where they're missing
- MD026 (no-trailing-punctuation): strips trailing `:` `!` `?` from
  ATX headings

Why this exists (Aaron 2026-04-26):
    "in python shape should be a queue that we are missing substraight
     primitives"

I'd been carrying this fix as `/tmp/md_fix.py` and re-typing it
across multiple drain ticks. Per Otto-346 principle (recurring
dynamic Python = signal a substrate primitive is missing), the
right home is `tools/hygiene/` checked into substrate. This tool
is the formalized version of the recurring pattern.

Usage:
    python3 tools/hygiene/fix-markdown-md032-md026.py FILE [FILE ...]
    python3 tools/hygiene/fix-markdown-md032-md026.py --dry-run FILE

Always idempotent: running on already-clean file is no-op.

Composes with:
- markdownlint-cli2 (.github/workflows/gate.yml lint-markdown job)
  — this tool produces input the linter accepts; the linter is the
  detection check
- Otto-341 (mechanism over discipline; markdownlint discipline
  becomes mechanism via this tool)
- Otto-346 candidate (recurring dynamic = missing primitive;
  this tool IS the primitive that absorbs the recurring pattern)
"""

import argparse
import re
import sys
from pathlib import Path


# All four CommonMark unordered list markers (`-`, `*`, `+`) plus
# ordered (`\d+\.`). markdownlint MD004 is disabled in this repo, so
# alternate markers do appear in committed files.
_LIST_LINE = re.compile(r"^(  )*([-*+] |\d+\. )")
_INDENTED_LINE = re.compile(r"^  +\S")

# MD026 strips trailing punctuation from ATX headings. markdownlint's
# default `punctuation` setting is `.,;:!?` (excluding `?` is configurable
# but we strip it here since the original tool stripped it). Allow
# optional trailing whitespace after the punctuation so headings like
# `## Title:   ` are still cleaned (the original regex required EOL
# immediately after the punctuation).
_HEADING_WITH_PUNCT = re.compile(r"^(#+ .+?)([.,;:!?]+)\s*$")

# Fenced-code-block delimiters. CommonMark allows ``` or ~~~ (3+ chars)
# at the start of a line (with optional info string after). Tilde and
# backtick fences cannot interrupt each other — track which fence opened.
_FENCE_OPEN = re.compile(r"^( {0,3})(`{3,}|~{3,})\s*([^`]*)$")

# YAML frontmatter line discriminator: matches a non-empty line that
# starts with a key followed by `:` (with optional leading whitespace
# and optional value). Examples that match: `id: B-0001`, `tags:`,
# `composes_with:`. YAML list items like `  - item` are intentionally
# NOT matched here; the discriminator only needs a simple `key:` check
# because the very-first content line of a real frontmatter should be
# a key:value, not a list item. List items + indented continuation
# lines are handled separately in the ratio check via the
# `_is_yaml_continuation` helper below.
_YAML_KEY_LINE = re.compile(r"^\s*[A-Za-z_][\w-]*\s*:")


def _is_yaml_continuation(line: str) -> bool:
    """Return True if a line plausibly continues a YAML key's value:
    indented continuation (>= 2 spaces or tab) of any non-blank
    content. Does NOT match column-0 lines (those should be either
    blank, a new key, or non-frontmatter content).

    Critically: this does NOT count `- item` at column 0 as
    continuation. A column-0 `- item` is almost always a markdown
    list, not a YAML list item under a parent key. YAML list items
    appear with leading indent (`  - X`); the leading indent is
    what disambiguates them.

    Used by the frontmatter-detection ratio check to count
    indented-continuation lines toward the "YAML-shaped majority"
    threshold without double-counting markdown list items as
    YAML."""
    if not line.strip():
        return False  # blank lines counted separately
    return line.startswith(("  ", "\t"))


def _is_list_or_continuation(line: str) -> bool:
    """Return True if line is a list item or its continuation
    (indented paragraph under a list item)."""
    return bool(_LIST_LINE.match(line) or _INDENTED_LINE.match(line))


def _is_list(line: str) -> bool:
    """Return True if line starts a list item."""
    return bool(_LIST_LINE.match(line))


def _classify_lines(lines: list[str]) -> list[bool]:
    """Return a boolean list `inside[i]` = True iff line `i` is inside
    a region that must NOT be touched by the MD032/MD026 transforms.

    Two such regions:

    1. **YAML frontmatter** (Jekyll/Hugo/factory-convention shape):
       file starts with a line `---`, line 1 is YAML-shaped
       (matches `key:` at start), and a closing `---` exists later.
       Lines from line 0 through the closing `---` are frontmatter.
       Inserting blanks here breaks YAML parsing (e.g.
       `composes_with:` followed by blank line then `  - X` parses
       as `composes_with: null` plus a separate top-level list).
       MD026 would only affect frontmatter if a YAML-key line
       happened to match the ATX-heading pattern (`^#+ `) — which
       it can't, since YAML keys don't start with `#`. So the YAML
       risk is concentrated in MD032's blank-insertion behavior,
       and the frontmatter-skip protects that.

    2. **Fenced code blocks**: a line starting with 3+ backticks or
       3+ tildes; closing fence must be the same character class as
       the opener and have at least as many characters. Inserting
       blanks here would mutate code examples (e.g. shell-script
       with `- option` flags would acquire spurious blanks).

    Both regions are conservatively treated as "inside" so transforms
    skip them. Better to skip than to corrupt structure.

    YAML frontmatter detection is conservative — must distinguish
    real frontmatter from a markdown file that happens to start with
    a thematic break (`---` followed by content):

      Real frontmatter:  line 0 is `---`,
                         line 1 looks YAML-shaped (`key: value` or `key:`),
                         a closing `---` exists later.
      Thematic break:    line 0 is `---`,
                         line 1 is markdown body (heading / prose / etc.).

    The YAML-shape check on line 1 is the discriminator. Without it,
    a file starting with a horizontal rule would have all subsequent
    content marked as "inside frontmatter," skipping every list and
    heading from being processed.

    Files without frontmatter (line 0 not `---`, or line 1 not
    YAML-shaped, or no closing `---`) skip the frontmatter region
    entirely — pass-through to the fence-detection logic."""
    inside: list[bool] = [False] * len(lines)

    # Pass 1: YAML frontmatter region — only if all FIVE conditions:
    # (a) line 0 is exactly `---`
    # (b) line 1 is YAML-shaped (matches `key:` at start, ignoring
    #     leading whitespace)
    # (c) a closing `---` line exists within the next 200 lines
    # (d) the line immediately BEFORE the closing `---` is YAML-
    #     shaped, blank, or a YAML continuation (catches the case
    #     where a closing `---` is just a thematic break — its
    #     preceding line is markdown prose, not a YAML key)
    # (e) at least 75% of non-blank lines BETWEEN the bookends are
    #     YAML-shaped (catches files where a single YAML-looking
    #     line appears after a thematic break, with a closing
    #     thematic break further down — typical "tip:" / "note:" /
    #     "warning:" prose patterns)
    #
    # The (b)/(d)/(e) checks together distinguish real frontmatter
    # from prose where `note: ...` happens to follow a thematic
    # break with another thematic break later. (c)'s 200-line cap
    # is defense-in-depth: real frontmatter is rarely >50 lines.
    _MAX_FRONTMATTER = 200
    _YAML_RATIO_MIN = 0.75
    if (
        len(lines) >= 2
        and lines[0].rstrip() == "---"
        and _YAML_KEY_LINE.match(lines[1])
    ):
        fm_end = -1
        for j in range(2, min(len(lines), _MAX_FRONTMATTER + 1)):
            if lines[j].rstrip() == "---":
                fm_end = j
                break
        if fm_end > 0:
            # Check (d): line immediately before closing `---` is
            # YAML-shaped, blank, or YAML continuation. (No need
            # for `if fm_end > 0` ternary — already in the
            # `if fm_end > 0:` block.)
            prev_line = lines[fm_end - 1]
            prev_ok = (
                not prev_line.strip()
                or _YAML_KEY_LINE.match(prev_line)
                or _is_yaml_continuation(prev_line)
            )
            # Check (e): YAML-shaped majority of non-blank lines.
            # Counts: `key:` lines (via _YAML_KEY_LINE) and
            # indented-continuation lines (via _is_yaml_continuation).
            # Does NOT count column-0 `- item` lines — those are
            # almost always markdown list items, not YAML, and
            # counting them would mis-classify ordinary markdown
            # documents (e.g. a thematic break + "note: ..." +
            # bullet list) as frontmatter.
            yaml_count = 0
            non_blank = 0
            for j in range(1, fm_end):
                if lines[j].strip():
                    non_blank += 1
                    if (
                        _YAML_KEY_LINE.match(lines[j])
                        or _is_yaml_continuation(lines[j])
                    ):
                        yaml_count += 1
            ratio_ok = (
                non_blank == 0
                or (yaml_count / non_blank) >= _YAML_RATIO_MIN
            )
            if prev_ok and ratio_ok:
                for k in range(fm_end + 1):  # inclusive of closing `---`
                    inside[k] = True
        # If conditions don't all hold, conservatively don't mark
        # any lines as frontmatter (treat as thematic break / body).

    # Pass 2: fenced code blocks (skip lines already marked
    # inside-frontmatter — they don't open / close fences).
    open_char: str | None = None  # '`' or '~'
    open_len: int = 0
    for i, line in enumerate(lines):
        if inside[i]:
            continue  # Already marked as frontmatter
        m = _FENCE_OPEN.match(line)
        if m and open_char is None:
            # Opening fence
            fence = m.group(2)
            open_char = fence[0]
            open_len = len(fence)
            inside[i] = True
        elif m and open_char is not None:
            # Possible closing fence — must be same char class and
            # length >= open_len, with no info string.
            fence = m.group(2)
            if fence[0] == open_char and len(fence) >= open_len and not m.group(3).strip():
                inside[i] = True  # The closing fence line itself
                open_char = None
                open_len = 0
            else:
                # A different fence char or shorter — still inside the
                # outer block (it's just code that looks fence-shaped).
                inside[i] = True
        else:
            inside[i] = open_char is not None
    return inside


def fix_md032(text: str) -> str:
    """Insert blank lines before list blocks (where the previous
    line is non-blank and not itself a list/continuation) and after
    list blocks (where the next line is non-blank and not part of
    the list).

    Skips lines inside fenced code blocks — inserting blanks there
    would mutate code examples (e.g. shell-script with `- option`
    flags would acquire spurious blanks)."""
    lines = text.split("\n")
    inside = _classify_lines(lines)

    # Pass 1: insert blank line BEFORE a list when previous is a
    # non-list, non-blank line. The boolean state needs to survive the
    # output mutation, so we map indices via the input position.
    out: list[str] = []
    out_inside: list[bool] = []
    for i, line in enumerate(lines):
        if not inside[i] and _is_list(line) and out:
            prev = out[-1]
            prev_inside = out_inside[-1]
            if prev.strip() and not prev_inside and not _is_list_or_continuation(prev):
                out.append("")
                out_inside.append(False)
        out.append(line)
        out_inside.append(inside[i])

    # Pass 2: insert blank line AFTER a list-item when next is a
    # non-list, non-blank line.
    out2: list[str] = []
    for i, line in enumerate(out):
        out2.append(line)
        if not out_inside[i] and _is_list(line) and i + 1 < len(out):
            nxt = out[i + 1]
            nxt_inside = out_inside[i + 1]
            if nxt.strip() and not nxt_inside and not _is_list_or_continuation(nxt):
                out2.append("")

    return "\n".join(out2)


def fix_md026(text: str) -> str:
    """Strip trailing `.` `,` `;` `:` `!` `?` punctuation (with optional
    trailing whitespace) from ATX heading lines (matches `^#+ ...`).

    Skips lines inside fenced code blocks — `# heading-shaped` lines
    inside code are content, not headings."""
    lines = text.split("\n")
    inside = _classify_lines(lines)
    out: list[str] = []
    for i, line in enumerate(lines):
        if inside[i]:
            out.append(line)
            continue
        m = _HEADING_WITH_PUNCT.match(line)
        if m:
            out.append(m.group(1))
        else:
            out.append(line)
    return "\n".join(out)


class FileNotFoundForFix(Exception):
    """Raised when an input file is missing — distinguishes from
    'no changes needed' so main() can exit non-zero."""


def fix_file(path: Path, dry_run: bool = False) -> tuple[bool, int]:
    """Apply both fixes to a file. Returns (changed, bytes_diff).

    Raises FileNotFoundForFix if the path does not exist, so the
    caller can distinguish missing-file (real error, exit non-zero)
    from clean-no-op (silent success)."""
    if not path.exists():
        raise FileNotFoundForFix(str(path))
    original = path.read_text()
    fixed = fix_md026(fix_md032(original))
    if fixed == original:
        return False, 0
    if not dry_run:
        path.write_text(fixed)
    return True, len(fixed) - len(original)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Fix markdownlint MD032 + MD026 violations mechanically"
    )
    parser.add_argument("files", nargs="+", help="Markdown files to fix")
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Print what would change; do not modify files",
    )
    args = parser.parse_args(argv)

    any_changed = False
    any_error = False
    for f in args.files:
        path = Path(f)
        try:
            changed, byte_diff = fix_file(path, dry_run=args.dry_run)
        except FileNotFoundForFix as exc:
            print(f"ERROR: file not found: {exc}", file=sys.stderr)
            any_error = True
            continue
        if changed:
            any_changed = True
            verb = "WOULD FIX" if args.dry_run else "FIXED"
            sign = "+" if byte_diff >= 0 else ""
            print(f"{verb} {path} ({sign}{byte_diff} bytes)")

    if any_error:
        # Suppress the misleading OK message when any input failed.
        return 1
    if not any_changed:
        print("OK: no changes needed")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/hygiene/fix-markdown-md032-md026.ts
#!/usr/bin/env bun
// fix-markdown-md032-md026.ts — mechanical fix for two markdownlint
// violations:
//
// - MD032 (blanks-around-lists): inserts blank lines before/after list
//   blocks where they're missing
// - MD026 (no-trailing-punctuation): strips trailing `.,;:!?` from
//   ATX headings
//
// Language choice: bun + TypeScript per
// `docs/DECISIONS/2026-04-20-tools-scripting-language.md`. Replaces the
// prior Python implementation per the 2026-04-29 directive to
// canonicalize on TS+Bun for tooling. Behavioural equivalence with
// the (now-retired) Python source is verified via the protocol in
// B-0086.
//
// Why this exists (per maintainer 2026-04-26):
//     "in python shape should be a queue that we are missing substraight
//      primitives"
//
// Per Otto-346 principle (recurring dynamic Python = signal a substrate
// primitive is missing), the right home is `tools/hygiene/` checked
// into substrate. This tool is the formalized version of the recurring
// pattern.
//
// Usage:
//     bun tools/hygiene/fix-markdown-md032-md026.ts FILE [FILE ...]
//     bun tools/hygiene/fix-markdown-md032-md026.ts --dry-run FILE
//
// Always idempotent: running on already-clean file is no-op.
//
// Composes with:
// - markdownlint-cli2 (.github/workflows/gate.yml lint-markdown job)
//   — this tool produces input the linter accepts; the linter is the
//   detection check
// - Otto-341 (mechanism over discipline; markdownlint discipline
//   becomes mechanism via this tool)
// - Otto-346 candidate (recurring dynamic = missing primitive;
//   this tool IS the primitive that absorbs the recurring pattern)

import { readFile, writeFile } from "node:fs/promises";

// CommonMark unordered list markers (`-`, `*`, `+`) plus ordered
// (`\d+\.`). markdownlint MD004 is disabled in this repo, so alternate
// markers do appear in committed files.
const LIST_LINE = /^( {2})*(?:[-*+] |\d+\. )/;
const INDENTED_LINE = /^ {2,}\S/;

// ATX heading test. Trailing-punctuation stripping is done procedurally
// in stripHeadingPunctuation — no regex for the punctuation match. The
// procedural form (a) avoids ReDoS shapes (sonarjs/slow-regex) and
// (b) has no length bound, matching Python's `[.,;:!?]+$` behavior on
// any input including machine-generated headings with arbitrarily long
// punctuation runs (PR #849 review noted that an earlier 10-char bound
// regressed equivalence on long runs).
const HEADING_LINE = /^#+ /;
const HEADING_PUNCT_CHARS = ".,;:!?";

// Fenced-code-block opening delimiter: `^( {0,3})(`{3,}|~{3,})...`.
// We capture only the indent + fence chars; whatever follows is the
// info string, which we slice out separately to avoid a regex shape
// sonarjs flags as ReDoS-adjacent.
const FENCE_DELIM = /^( {0,3})(`{3,}|~{3,})/;

// YAML frontmatter line discriminator: matches a non-empty line that
// starts with a key followed by colon (with optional leading whitespace
// and optional value).
const YAML_KEY_LINE = /^\s*[A-Za-z_][\w-]*\s*:/;

const MAX_FRONTMATTER = 200;
const YAML_RATIO_MIN = 0.75;

// Discriminated-union result for fixFile — Result-style for the
// expected outcomes (`not-found` / `unchanged` / `fixed`) per Zeta's
// Result-over-exception convention (CLAUDE.md ground rules).
// Unexpected I/O errors (non-ENOENT read, write failures) still
// propagate as exceptions for the caller to convert to an exit code;
// only the expected-outcome surface is exception-free. The
// `bytesDiff` is the UTF-16 code-unit delta; for ASCII content it
// matches Python's `len(fixed) - len(original)`.
type FixResult =
  | { kind: "unchanged" }
  | { kind: "fixed"; bytesDiff: number }
  | { kind: "not-found" };

interface FenceState {
  openChar: string | null;
  openLen: number;
}

interface BlanksBeforePass {
  out: string[];
  outInside: boolean[];
}

interface Args {
  files: string[];
  dryRun: boolean;
}

// Strip trailing punctuation (with optional trailing whitespace) from
// an ATX heading line. Returns the cleaned line, or null if the line is
// not a heading or has no trailing punctuation. Procedural form avoids
// the lazy-quantifier shape sonarjs flags.
function stripHeadingPunctuation(line: string): string | null {
  if (!HEADING_LINE.test(line)) return null;
  const trimmed = line.trimEnd();
  // Find where the heading prefix ends (after `#+ `). Used to preserve
  // at least ONE character of heading text below — see floor comment.
  const prefixMatch = HEADING_LINE.exec(trimmed);
  const prefixEnd = prefixMatch ? prefixMatch[0].length : 0;
  // Walk backwards from end-of-string, counting trailing chars that
  // belong to the punctuation set. Procedural form (no regex) so:
  //   (a) no ReDoS shape (sonarjs/slow-regex stays clean)
  //   (b) no length bound (matches Python's `[.,;:!?]+$` on any input
  //       length, including machine-generated headings with arbitrarily
  //       long punctuation runs)
  //   (c) preserves at least one character of heading text — Python's
  //       `_HEADING_WITH_PUNCT` regex used `.+?` before the punctuation
  //       class, requiring at least one non-punct char between `#+ `
  //       and the trailing punctuation run. A punctuation-only heading
  //       like `# !!!` should keep one char (`# !!`), not be erased
  //       to `# `. The `i > prefixEnd + 1` floor enforces this.
  let i = trimmed.length;
  while (i > prefixEnd + 1 && HEADING_PUNCT_CHARS.includes(trimmed[i - 1] ?? "")) {
    i -= 1;
  }
  if (i === trimmed.length) return null; // no trailing punctuation
  return trimmed.slice(0, i);
}

// Return true if a line plausibly continues a YAML key's value:
// indented continuation (>= 2 spaces or tab) of any non-blank content.
// Does NOT match column-0 lines.
function isYamlContinuation(line: string): boolean {
  if (line.trim() === "") return false;
  return line.startsWith("  ") || line.startsWith("\t");
}

function isYamlOk(line: string): boolean {
  return (
    line.trim() === "" ||
    YAML_KEY_LINE.test(line) ||
    isYamlContinuation(line)
  );
}

function isList(line: string): boolean {
  return LIST_LINE.test(line);
}

function isListOrContinuation(line: string): boolean {
  return LIST_LINE.test(line) || INDENTED_LINE.test(line);
}

// Locate the closing `---` for a candidate YAML frontmatter block, or
// -1 if no closing fence appears within MAX_FRONTMATTER lines.
function findFrontmatterEnd(lines: readonly string[]): number {
  const upper = Math.min(lines.length, MAX_FRONTMATTER + 1);
  for (let j = 2; j < upper; j += 1) {
    if (lines[j]?.trimEnd() === "---") return j;
  }
  return -1;
}

// Compute the YAML-shaped ratio of non-blank lines between the bookends
// (frontmatter check (e)). Returns true if the ratio meets the minimum
// or if there are no non-blank lines (vacuously true).
function frontmatterRatioOk(lines: readonly string[], fmEnd: number): boolean {
  let yamlCount = 0;
  let nonBlank = 0;
  for (let j = 1; j < fmEnd; j += 1) {
    const lj = lines[j];
    if (lj === undefined || lj.trim() === "") continue;
    nonBlank += 1;
    if (YAML_KEY_LINE.test(lj) || isYamlContinuation(lj)) yamlCount += 1;
  }
  return nonBlank === 0 || yamlCount / nonBlank >= YAML_RATIO_MIN;
}

// YAML frontmatter region — only if all FIVE conditions hold:
// (a) line 0 is exactly `---` (after trailing-whitespace strip)
// (b) line 1 is YAML-shaped (matches `key:` at start, ignoring leading
//     whitespace)
// (c) a closing `---` line exists within the next 200 lines
// (d) the line immediately BEFORE the closing `---` is YAML-shaped,
//     blank, or a YAML continuation
// (e) at least 75% of non-blank lines BETWEEN the bookends are
//     YAML-shaped
//
// Returns -1 when no frontmatter is present, otherwise returns the
// index of the closing `---`.
function findFrontmatterRegion(lines: readonly string[]): number {
  if (lines.length < 2) return -1;
  const line0 = lines[0];
  const line1 = lines[1];
  if (line0 === undefined || line1 === undefined) return -1;
  if (line0.trimEnd() !== "---") return -1;
  if (!YAML_KEY_LINE.test(line1)) return -1;
  const fmEnd = findFrontmatterEnd(lines);
  if (fmEnd <= 0) return -1;
  if (!isYamlOk(lines[fmEnd - 1] ?? "")) return -1;
  if (!frontmatterRatioOk(lines, fmEnd)) return -1;
  return fmEnd;
}

// Parse a fence-delimiter line into { fenceChar, fenceLen, infoString }
// or null if the line is not a fence. Procedural to keep regex bounded;
// the info string is whatever follows the fence chars.
function parseFenceLine(line: string): {
  fenceChar: string;
  fenceLen: number;
  info: string;
} | null {
  const m = FENCE_DELIM.exec(line);
  if (m === null) return null;
  const fence = m[2] ?? "";
  const fenceChar = fence[0] ?? "";
  const fenceLen = fence.length;
  const tail = line.slice((m[1]?.length ?? 0) + fenceLen).trimStart();
  // The Python source enforced `([^`]*)$` on the info string for BOTH
  // backtick and tilde fences — a backtick anywhere in the info string
  // invalidates the fence delimiter regardless of fence char. We honour
  // the same constraint procedurally for behavioural parity with the
  // retired Python tool (`fix-markdown-md032-md026.py`).
  if (tail.includes("`")) return null;
  return { fenceChar, fenceLen, info: tail };
}

// Update fence state and the `inside` mark for a single line.
function processFenceLine(
  line: string,
  i: number,
  inside: boolean[],
  state: FenceState,
): void {
  const parsed = parseFenceLine(line);
  if (parsed === null) {
    inside[i] = state.openChar !== null;
    return;
  }
  if (state.openChar === null) {
    state.openChar = parsed.fenceChar;
    state.openLen = parsed.fenceLen;
    inside[i] = true;
    return;
  }
  // Possible closing fence — must be same char class, length >=
  // openLen, and have no info string.
  const isClose =
    parsed.fenceChar === state.openChar &&
    parsed.fenceLen >= state.openLen &&
    parsed.info === "";
  inside[i] = true; // both closing line and lookalike are inside
  if (isClose) {
    state.openChar = null;
    state.openLen = 0;
  }
}

// Return a boolean array `inside[i]` = true iff line `i` is inside a
// region that must NOT be touched by the MD032/MD026 transforms.
//
// Two such regions: YAML frontmatter (5-condition discriminator) and
// fenced code blocks (matching open/close char + length).
function classifyLines(lines: readonly string[]): boolean[] {
  const inside = new Array<boolean>(lines.length).fill(false);

  const fmEnd = findFrontmatterRegion(lines);
  if (fmEnd > 0) {
    for (let k = 0; k <= fmEnd; k += 1) inside[k] = true;
  }

  // Pass 2: fenced code blocks (skip lines already marked inside
  // frontmatter — they don't open / close fences).
  const state: FenceState = { openChar: null, openLen: 0 };
  for (let i = 0; i < lines.length; i += 1) {
    if (inside[i]) continue;
    const line = lines[i];
    if (line === undefined) continue;
    processFenceLine(line, i, inside, state);
  }
  return inside;
}

// Insert blank lines before list blocks (where the previous line is
// non-blank and not itself a list/continuation).
function insertBlanksBefore(
  lines: readonly string[],
  inside: readonly boolean[],
): BlanksBeforePass {
  const out: string[] = [];
  const outInside: boolean[] = [];
  for (let i = 0; i < lines.length; i += 1) {
    const line = lines[i];
    if (line === undefined) continue;
    const prev = out.at(-1);
    const prevInside = outInside.at(-1) ?? false;
    if (
      !inside[i] &&
      isList(line) &&
      prev !== undefined &&
      prev.trim() !== "" &&
      !prevInside &&
      !isListOrContinuation(prev)
    ) {
      out.push("");
      outInside.push(false);
    }
    out.push(line);
    outInside.push(inside[i] ?? false);
  }
  return { out, outInside };
}

// Insert blank lines after list blocks (where the next line is
// non-blank and not part of the list).
function insertBlanksAfter(
  lines: readonly string[],
  inside: readonly boolean[],
): string[] {
  const out: string[] = [];
  for (let i = 0; i < lines.length; i += 1) {
    const line = lines[i];
    if (line === undefined) continue;
    out.push(line);
    const nxt = lines[i + 1];
    const nxtInside = inside[i + 1] ?? false;
    if (
      !inside[i] &&
      isList(line) &&
      nxt !== undefined &&
      nxt.trim() !== "" &&
      !nxtInside &&
      !isListOrContinuation(nxt)
    ) {
      out.push("");
    }
  }
  return out;
}

// Insert blank lines around list blocks. Skips lines inside fenced code
// blocks — inserting blanks there would mutate code examples (e.g.
// shell-script with `- option` flags would acquire spurious blanks).
function fixMd032(text: string): string {
  const lines = text.split("\n");
  const inside = classifyLines(lines);
  const before = insertBlanksBefore(lines, inside);
  return insertBlanksAfter(before.out, before.outInside).join("\n");
}

// Strip trailing punctuation (with optional trailing whitespace) from
// ATX heading lines. Skips lines inside fenced code blocks.
function fixMd026(text: string): string {
  const lines = text.split("\n");
  const inside = classifyLines(lines);
  return lines
    .map((line, i) => {
      if (inside[i]) return line;
      return stripHeadingPunctuation(line) ?? line;
    })
    .join("\n");
}

// Apply both fixes to a file. Returns a discriminated-union result.
//
// Uses try-readFile + catch ENOENT (instead of existsSync-then-readFile)
// to avoid TOCTOU race-condition (CWE-367, CodeQL flagged). The
// existsSync pattern leaves a race window where the file is deleted
// between the check and the read; catching ENOENT on the read itself
// is atomic at the OS-syscall level.
async function fixFile(path: string, dryRun: boolean): Promise<FixResult> {
  let original: string;
  try {
    original = await readFile(path, "utf8");
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") {
      return { kind: "not-found" };
    }
    throw err;
  }
  const fixed = fixMd026(fixMd032(original));
  if (fixed === original) return { kind: "unchanged" };
  if (!dryRun) await writeFile(path, fixed);
  return { kind: "fixed", bytesDiff: fixed.length - original.length };
}

function fail(message: string, code: number): never {
  process.stderr.write(`ERROR: ${message}\n`);
  process.exit(code);
}

function parseArgs(argv: readonly string[]): Args {
  const args: Args = { files: [], dryRun: false };
  for (const a of argv) {
    if (a === "--dry-run") {
      args.dryRun = true;
    } else if (a === "-h" || a === "--help") {
      printHelp();
      process.exit(0);
    } else if (a.startsWith("--")) {
      fail(`unknown argument: ${a}`, 2);
    } else {
      args.files.push(a);
    }
  }
  if (args.files.length === 0) fail("at least one FILE is required", 2);
  return args;
}

function printHelp(): void {
  console.log(
    [
      "fix-markdown-md032-md026.ts — mechanical MD032/MD026 fix",
      "",
      "Usage:",
      "  bun tools/hygiene/fix-markdown-md032-md026.ts FILE [FILE ...]",
      "  bun tools/hygiene/fix-markdown-md032-md026.ts --dry-run FILE",
      "",
      "Always idempotent: running on already-clean file is no-op.",
    ].join("\n"),
  );
}

async function processOneFile(
  path: string,
  dryRun: boolean,
): Promise<{ changed: boolean; error: boolean }> {
  const result = await fixFile(path, dryRun);
  switch (result.kind) {
    case "not-found":
      process.stderr.write(`ERROR: file not found: ${path}\n`);
      return { changed: false, error: true };
    case "unchanged":
      return { changed: false, error: false };
    case "fixed": {
      const verb = dryRun ? "WOULD FIX" : "FIXED";
      const sign = result.bytesDiff >= 0 ? "+" : "";
      // Explicit String() coercion satisfies @typescript-eslint/restrict-
      // template-expressions (configured strict — won't accept number in
      // template literal without explicit conversion). Output equivalence
      // with Python's `+{n}` / `{n}` formatting is preserved.
      console.log(`${verb} ${path} (${sign}${String(result.bytesDiff)} bytes)`);
      return { changed: true, error: false };
    }
  }
}

async function main(): Promise<number> {
  const args = parseArgs(process.argv.slice(2));

  let anyChanged = false;
  let anyError = false;
  for (const f of args.files) {
    const { changed, error } = await processOneFile(f, args.dryRun);
    if (changed) anyChanged = true;
    if (error) anyError = true;
  }

  if (anyError) return 1;
  if (!anyChanged) console.log("OK: no changes needed");
  return 0;
}

// Set process.exitCode (instead of process.exit) so pending stdout/
// stderr writes flush before the process terminates. Mirrors the
// tally.ts pattern.
process.exitCode = await main();
